dmpr/message.py

```diff
@@ -0,0 +1,193 @@
+
+        if msg['type'] == 'full':
+            # Check for at least one address
+            if not any(('addr-v4' in msg, 'addr-v6' in msg)):
+                raise InvalidMessage("Need one of add addr-v4 and addr-v6 in "
+                                     "message")
+
+        # Partial messages need a base
+        elif msg['type'] == 'partial':
+            if 'partial-base' not in msg:
+                raise InvalidMessage("Need a partial-base in partial updates")
+            if msg['partial-base'] != self._base.seq:
+                raise InvalidPartialUpdate("Wrong base message sequence number")
+
+        # If routing-data exists and there is at least one policy with paths
+        # then there must be link-attributes
+        # Also every policy value must be a dict
+        routing_data = msg.get('routing-data', {})
+        if routing_data and any(routing_data.values()):
+            if not msg.get('link-attributes'):
+                raise InvalidMessage("Need link-attributes for routing-data")
+            for i in routing_data.values():
+                if not isinstance(i, dict):
+                    raise InvalidMessage("routing-data policies must be dict")
+
+        for i in ('routing-data', 'node-data', 'link-attributes'):
+            if i in msg:
+                if not isinstance(msg[i], dict):
+                    raise InvalidMessage("{} must be dict".format(i))
+
+    def _apply_full(self, msg: dict) -> bool:
+        update = False
+        self.seq = msg['seq']
+        self.addr_v4 = msg.get('addr-v4', None)
+        self.addr_v6 = msg.get('addr-v6', None)
+
+        update |= self._save_networks(msg)
+
+        msg_routing_data = msg.get('routing-data', {})
+        msg_link_attributes = msg.get('link-attributes', {})
+        msg_node_data = msg.get('node-data', {})
+        new_routing_data = {}
+        for policy in msg_routing_data:
+            new_routing_data[policy] = {}
+            for node in msg_routing_data[policy]:
+                path = self._get_path(msg_routing_data[policy][node]['path'],
+                                      msg_link_attributes)
+                if not path:
+                    # Loop detection
+                    continue
+
+                new_routing_data[policy][node] = {
+                    'path': path,
+                }
+
+                if node not in msg_node_data:
+                    # We have no node_data for this node, apparently it does not
+                    # advertise any networks
+                    continue
+
+        if self.routing_data != new_routing_data:
+            update = True
+            self.routing_data = new_routing_data
+
+        if self.node_data != msg_node_data:
+            update = True
+            self.node_data = msg_node_data
+
+        self._base = BaseMessage()
+        self._base.apply_base_data(self)
+
+        return update
+
+    def _apply_partial(self, msg: dict) -> bool:
```

Comment:
sometimes we need more comments. The code is clear but it is still not *that* easy to know what the functions does and what particular loops do. Some small comments before functional blocks are helpful too, beside a general comment at the top of the function what a function in general do. Right?

Reply:
Yeah, comments need to be improved, I will work on that.